### PR TITLE
Cache HTTP routes

### LINKED ISSUE #488: Improve routing speed
[Prism](https://github.com/vladfaust/prism) uses @luislavena's Radix tree implementation as well. I've improved route resolve times by using `Hash` for static routes lookup as seen @ https://github.com/vladfaust/prism/issues/45.

You can easily improve Kemal's speed as well, see https://github.com/vladfaust/prism/blob/46e9ec7a214e1976ddc88f465f3c0189e5557c33/src/prism/router.cr#L182-L212. I'm open for questions, if any.

Cheers.

### PR DESCRIPTION
**Pull request by sdogruyol**
### Description of the Change

Fixes #488 

### Benefits

%20 - %25 overall performance🚀 

### Possible Drawbacks

None

### CODE REVIEW

**Comment by straight-shoota** [comment]
A possible drawback is memory flooding. All matched path are cached in the hash forever. This can be bad if you have a lot of unique paths (think of one-time tokens). These shouldn't be cached or at least released from cache at some point. Spamming a kemal app with valid paths could be a potential attack vector.

A cache should always have some limitation to keep it from growing uncontrolled. You can't control the cache keys because the requested paths are user generated. So there should be a size limit for the cache, age limit for cache entries or some other metric (or combinations of those).

**Comment by sdogruyol** [comment]
@straight-shoota I agree that there should be a limit on the cache 👍 

I believe a cache size limit would be a good choice for starters and can be improved later 👍

File: src/kemal/route_handler.cr [hunk 1]
```diff
@@ -33,7 +34,11 @@ module Kemal
 
       route = @routes.find(lookup_path)
 
-      @cached_routes[lookup_path] = route if route.found?
+      if route.found? && @cached_routes.size < CACHED_ROUTES_LIMIT
+        @cached_routes[lookup_path] = route
+      else
```
Comment (straight-shoota) [inline src/kemal/route_handler.cr]: This clears the cache every time a path doesn't resolve to a route. Not very effective.

**Comment by straight-shoota** [comment]
Clearing the cache completely once it is full doesn't sound very smart, but I guess it's good enough for now and can be improved later.

**Comment by ohenepee** [comment]
How about implementing a stack where every time a route is used/accessed, it is placed at the top of the stack; thus when the `cached_routes` is full, we just pop the last route at the bottom of the stack, which will probably be the oldest and least used route. This way there'll be no need to clear the entire `cached_routes` when its full. @sdogruyol

**Comment by nhh** [comment]
I dont think caching the http routes is a good practice to increase performance.

Consider the case where your app is providing a ton of path variables, looking them up on a cache, that is limited will generate overhead. And in the worst cade will lead to many unnessesary cache invalidations.

Caching is hard, maybe compiling a regex within the patch matcher or something would fit better.

Btw 20% Peformance sounds a lot, but thats probably  20% of 1-2ms.

What do you think?

**Pull request merged**